sequencer/src/block/namespace_payload/ns_payload.rs

```diff
@@ -0,0 +1,137 @@
+use crate::{
+    block::namespace_payload::types::{
+        FromNsPayloadBytes, NsPayloadByteLen, NsPayloadBytesRange, NumTxs, NumTxsRange,
+        NumTxsUnchecked, TxIndex, TxIter, TxPayloadRange, TxTableEntriesRange,
+    },
+    NamespaceId, Transaction,
+};
+use serde::{Deserialize, Serialize};
+
+/// Raw binary data for a single namespace's payload.
+///
+/// Any sequence of bytes is a valid [`NsPayload`].
+///
+/// See module-level documentation [`types`](super::types) for a full
+/// specification of the binary format of a namespace.
+pub(in crate::block) struct NsPayload([u8]);
+
+impl NsPayload {
+    pub fn from_bytes_slice(bytes: &[u8]) -> &NsPayload {
+        NsPayload::new_private(bytes)
+    }
+    pub fn as_bytes_slice(&self) -> &[u8] {
+        &self.0
+    }
+    pub fn byte_len(&self) -> NsPayloadByteLen {
+        NsPayloadByteLen::from_usize(self.0.len())
+    }
+
+    /// Read and parse bytes from the ns payload.
+    ///
+    /// Arg `range: &R` is convertible into a `Range<usize>` via
+    /// [`NsPayloadBytesRange`]. The payload bytes are parsed into a `R::Output`
+    /// via [`FromNsPayloadBytes`].
+    pub fn read<'a, R>(&'a self, range: &R) -> R::Output
+    where
+        R: NsPayloadBytesRange<'a>,
+    {
+        <R::Output as FromNsPayloadBytes<'a>>::from_payload_bytes(&self.0[range.ns_payload_range()])
+    }
+
+    /// Iterator over all transactions in this namespace.
+    pub fn iter(&self) -> TxIter {
+        self.iter_from_num_txs(&self.read_num_txs())
+    }
+
+    /// Return all transactions in this namespace. The namespace ID for each
+    /// returned [`Transaction`] is set to `ns_id`.
+    pub fn export_all_txs(&self, ns_id: &NamespaceId) -> Vec<Transaction> {
```

Comment:
nit: (just a comment) providing `ns_id` seems redundant in many cases (NsPayload should already be aware of the corresponding NsId). Extending NsPayload with a NsId (making NsPayload id-aware) would fix this but likely messes up the desired abstraction.

Reply:
I wrestled with this question, and I even have a id-aware implementation in a draft branch: https://github.com/EspressoSystems/espresso-sequencer/blob/5feebba5a351ab8152a72f5a8478a63f82da362f/sequencer/src/block2/ns_payload.rs#L53-L58

TBH I can't quite remember why I decided against it. Your suggestion that it "messes up the desired abstraction" sounds familiar though. Some quick notes:
- Some computation would be repeated or unnecessary, especially in `TxProof`.
- If `NsPayload` has any other fields beyond `[u8]` then it breaks `NsPayloadOwned` due to limitations in Rust, which would force me to write a lot of boilerplate code for delegation, etc.

I agree it's not ideal to have a `ns_id` arg here. I'm willing to reconsider this question, but I also want to be mindful not to spend time chasing rabbits down holes. 😉 🐰 🕳️